app/code/Magento/Newsletter/Controller/Subscriber/NewAction.php

```diff
@@ -125,29 +132,38 @@ public function execute()
                 $this->validateEmailAvailable($email);
 
                 $subscriber = $this->_subscriberFactory->create()->loadByEmail($email);
-                if ($subscriber->getId()
-                    && $subscriber->getSubscriberStatus() == \Magento\Newsletter\Model\Subscriber::STATUS_SUBSCRIBED
-                ) {
-                    throw new \Magento\Framework\Exception\LocalizedException(
+                if ($subscriber->getId() && $subscriber->getSubscriberStatus() == Subscriber::STATUS_SUBSCRIBED) {
+                    throw new LocalizedException(
                         __('This email address is already subscribed.')
                     );
                 }
 
-                $status = $this->_subscriberFactory->create()->subscribe($email);
-                if ($status == \Magento\Newsletter\Model\Subscriber::STATUS_NOT_ACTIVE) {
-                    $this->messageManager->addSuccess(__('The confirmation request has been sent.'));
-                } else {
-                    $this->messageManager->addSuccess(__('Thank you for your subscription.'));
-                }
-            } catch (\Magento\Framework\Exception\LocalizedException $e) {
-                $this->messageManager->addException(
+                $this->messageManager->addSuccessMessage($this->getSuccessMessage(
```

Comment:
@ihor-sviziev I refactored it slightly different, using a private function instead of the if/else block

Reply:
@ihor-sviziev  I was indeed doubting about this one, the old way is indeed more understandable. :+1: